Graceful restart for EVPN

Pdoijode opened the pull request:
This MR contains changes for EVPN GR and GR to support multihop BGP sessions. Please refer to individual commits.

ton31337 commented on bgpd/bgp_evpn.c:
```diff
@@ -5942,6 +5957,99 @@ void bgp_evpn_handle_autort_change(struct bgp *bgp)
+		return;
+
+	ctx->cnt = bgp_deferred_path_selection(bgp, afi, safi, vpn->mac_table, ctx->cnt, vpn, true);
+	ctx->cnt = bgp_deferred_path_selection(bgp, afi, safi, vpn->ip_table, ctx->cnt, vpn, true);
```
Is this correct really? Why do we "reset" cnt? Shouldn't it be a sum?

Pdoijode commented:
@ton31337 Graceful Restart RFC does not define multi-level "deferring", this is implementation-specific case.

mrin-g commented on lib/zebra.h:
```diff
@@ -189,9 +189,9 @@ typedef enum {
-		for (safi = SAFI_UNICAST; safi <= SAFI_MPLS_VPN; safi++)
+#define FOREACH_AFI_SAFI_NSF(afi, safi)                                                           \
+	for (afi = AFI_IP; afi < AFI_MAX; afi++)                                                  \
+		for (safi = SAFI_UNICAST; safi <= SAFI_EVPN; safi++)
```
Will this not enable GR for ENCAP SAFI as well ?

Pdoijode replied:
@mrin-g Mrinmoy, No, GR is now enabled for L2VPN EVPN AFI-SAFI in addition to v4 and v6 unicast.

kacpekwasny replied:
@mrin-g  I think there is another function controlling this:

https://github.com/FRRouting/frr/pull/19778/commits/7e2045c994152b3d08c1b470c6f77310fc184457#diff-f7b4da9b53d00be700be61d8bbd3d0113a456949a67b22b566abc6cd089657c5R3130

ton31337 reviewed (changes requested):
Topotest is required to verify GR for EVPN...

Pdoijode commented:
@ton31337 @riw777 I am working on adding topotest for l2vpn evpn GR and for GR with multihop BGP session

mrin-g commented on lib/zebra.h:
```diff
@@ -189,9 +189,9 @@ typedef enum {
-		for (safi = SAFI_UNICAST; safi <= SAFI_MPLS_VPN; safi++)
+#define FOREACH_AFI_SAFI_NSF(afi, safi)                                                           \
+	for (afi = AFI_IP; afi < AFI_MAX; afi++)                                                  \
+		for (safi = SAFI_UNICAST; safi <= SAFI_EVPN; safi++)
```
You may consider, to avoid missing out the bgp_gr_supported_for_afi_safi check and keeping it consolidted:
```suggestion
#define FOREACH_AFI_SAFI_NSF(afi, safi)                                                        \
	for (afi = AFI_IP; afi < AFI_MAX; afi++)                                                   \
		for (safi = SAFI_UNICAST; safi <= SAFI_EVPN; safi++)                                   \
		    if (bgp_gr_supported_for_afi_safi(afi,safi)
```

ton31337 commented:
This commit https://github.com/FRRouting/frr/pull/19778/commits/7d02858185b43debe44c432dbeea2d92219c8b9c is against the standards.

Pdoijode commented:
For [lib, zebra: Ensure time has enough accuracy on startup](https://github.com/FRRouting/frr/pull/19778/commits/376054b2a4e904effadddfa3b20d35d58b1350e7) there's already a MR open: https://github.com/FRRouting/frr/pull/12975. But without this commit, test_bgp_evpn_gr_select_deferral_cleanup_on_pe2 in tests/topotests/bgp_evpn_gr/test_bgp_evpn_gr.py fails. So keeping this commit temporarily here to get CI passing.

Pdoijode commented:
ci:rerun

ton31337 commented on tests/topotests/bgp_gr_multihop/r1/bgpd.conf:
```diff
@@ -0,0 +1,32 @@
+log syslog debugging
+log file /tmp/topotests/bgp_gr_multihop.test_bgp_gr_multihop/r1/bgpd.log
```
This must be dropped.

ton31337 commented on tests/topotests/bgp_evpn_gr/test_bgp_evpn_gr.py:
```diff
@@ -0,0 +1,752 @@
+
+    # Load FRR configuration and start routers
+    for name, router in tgen.routers().items():
+        router.load_config(TopoRouter.RD_ZEBRA, os.path.join(CWD, f"{name}/zebra.conf"))
```
Please use frr.conf (instead of per daemon configs).

Pdoijode replied:
Done

ton31337 replied:
Not done?

Pdoijode commented on tests/topotests/bgp_gr_multihop/r1/bgpd.conf:
```diff
@@ -0,0 +1,32 @@
+log file /tmp/topotests/bgp_gr_multihop.test_bgp_gr_multihop/r1/bgpd.log
+log timestamp precision 6
+debug bgp neighbor-events
+debug bgp graceful-restart
```
Done

Pdoijode commented on tests/topotests/bgp_gr_multihop/r1/bgpd.conf:
```diff
@@ -0,0 +1,32 @@
+log syslog debugging
+log file /tmp/topotests/bgp_gr_multihop.test_bgp_gr_multihop/r1/bgpd.log
+log timestamp precision 6
+debug bgp neighbor-events
```
Done

Pdoijode commented:
@ton31337 I have addressed your comments. Could you please take a look?
For 0e056bb2df75dd291a1072a7e8533d03e4b4b9ca (topotest for L2vpn GR) I tried combining bgpd.cong and zebra.conf into frr.conf but ran into issue! So I will keep those separate.

Pdoijode commented on tools/frr_babeltrace.py:
```diff
@@ -20,6 +20,79 @@
+    if field_val == 1:
+        return "Check all EORs"
+    elif field_val == 2:
+        return "All dir conn EORs rcvd"
```
I haven't checked if it warps.

Pdoijode commented:
@riw777 @ton31337 Is there anything else required for this MR? If not, could you please approve?

ton31337 commented on bgpd/bgp_trace.h:
```diff
@@ -636,6 +636,181 @@ TRACEPOINT_EVENT(
+	TP_ARGS(char *, bgp_name, uint8_t, afi, uint8_t, safi,
+		uint32_t, defer_time, uint8_t, loc),
+	TP_FIELDS(ctf_string(bgp_instance, bgp_name)
+		  ctf_integer(uint8_t, afi, afi)
```
Could you fix the alignment?

Pdoijode replied:
Fixed it

ton31337 reviewed (commented):
bgp_evpn_gr topotest is still using per-daemon configs, could you switch to frr.conf?

Pdoijode commented:
@ton31337  I have combined bgpd.conf and zebra.conf into frr.conf and fixed the test too. Could you please take a look?

Pdoijode commented:
ci:rerun
